Add iSCSI ReplacementTimeout  API's to keep volume in RW state during "Node Down" scenario

mittachaitu opened the pull request:
**What this PR does / why we need it**:
 Add ReplacementTimeout field in ISCSIVolumeSource and
 ISCSIPersistentVolumeSource spec as API changes.
 ReplacementTimeout controls how long the iSCSI layer should wait
 for a timed-out path/session to re-establish itself before failing
 any commands on it.

Below one is the example to pass a ReplacementTimeout setting parameter:
```yaml
apiVersion: v1
kind: PersistentVolume
metadata:
  name: iscsi-pv
spec:
  storageClassName: standard
  capacity:
    storage: 1Gi
  accessModes:
    - ReadWriteOnce
  iscsi:
     targetPortal: 192.68.0.134:3260
     iqn: iqn.2018-09.com.storage.volume:vol1
     lun: 0
     fsType: 'ext4'
     readOnly: false
     replacementTimeout: 300
 ```

cc /sig storage
```release-note
NONE
```


Signed-off-by: mittachaitu <sai.chaithanya@mayadata.io>

prateekpandey14 commented:
cc @rootfs

jsafrane commented:
> ```
> provisioner: kubernetes.io/storageos
> parameters:
>   storage.alpha.kubernetes.io/iscsiTimeout: "node.session.timeo.replacement_timeout=350"
> ```

That does not make sense, `kubernetes.io/storageos` provisioner will create StorageOS volumes, not iscsi.

jsafrane commented on pkg/volume/iscsi/iscsi.go:
```diff
@@ -53,6 +53,18 @@ const (
+	Name  iscsiTimeoutKey
+	Value string
+}
+type iscsiTimeouts []iscsiTimeout
```
Why is it array? There  can be only one annotation in a PV.

jsafrane commented on pkg/volume/iscsi/iscsi.go:
```diff
@@ -506,6 +519,26 @@ func getISCSISecretNameAndNamespace(spec *volume.Spec, defaultSecretNamespace st
 }
 
+// updateFromPV will read iSCSI Timeout settings from Persistent Volume object
+func (kvp iscsiTimeouts) getISCSITimeoutSettingInfo(spec *volume.Spec) iscsiTimeouts {
```
Please rewrite to `func getISCSITimeoutSettingInfo(...) iscsiTimeouts`, there is no need for this to be method of `iscsiTimeouts`

jsafrane commented on pkg/volume/iscsi/iscsi_util.go:
```diff
@@ -86,10 +88,34 @@ func updateISCSIDiscoverydb(b iscsiDiskMounter, tp string) error {
+	if b.iscsiTimeouts != nil {
+		for _, iscsiSetting := range b.iscsiTimeouts {
+			if iscsiSetting.Name != "" && iscsiSetting.Value != "" {
+				out, err := b.exec.Run("iscsiadm", "-m", "node", "-p", tp, "-T", b.Iqn, "-I", b.Iface, "-o", "update", "-n", string(iscsiSetting.Name), "-v", iscsiSetting.Value)
```
This just blindly sets **any** parameters from the annotation to `iscsiadm`. It does not check if it's timeout related. This should be validated somewhere and you should allow only very specific keys.

rootfs commented:
/cc @jsafrane

rootfs commented:
/assign @jsafrane

rootfs commented:
iscsi timeout can be set in the default iscsid.conf (on rhel/centos/fedora it is /etc/iscsi/iscsid.conf) on the nodes

mittachaitu commented:
> > ```
> > provisioner: kubernetes.io/storageos
> > parameters:
> >   storage.alpha.kubernetes.io/iscsiTimeout: "node.session.timeo.replacement_timeout=350"
> > ```
> 
> That does not make sense, `kubernetes.io/storageos` provisioner will create StorageOS volumes, not iscsi.

It's just an example, anyways updated to `kubernetes.io/provisioner_iscsi`

prateekpandey14 commented on pkg/volume/iscsi/iscsi.go:
```diff
@@ -53,6 +53,18 @@ const (
+type iscsiTimeouts []iscsiTimeout
+
+//iscsiAnnotationLblKey is used to read timeout setting and value from PV object
+const iscsiAnnotationLblKey = "storage.alpha.kubernetes.io/iscsiTimeout"
```
I was thinking the same, should we add `TimeOut` API field in `ISCSIVolumeSource` and validate the same ?

mittachaitu commented:
> iscsi timeout can be set in the default iscsid.conf (on rhel/centos/fedora it is /etc/iscsi/iscsid.conf) on the node

mittachaitu commented:
@jsafrane added the ReplacementTimeout as proper API field instead of annotation, PTAL

jsafrane commented on pkg/apis/core/validation/validation_test.go:
```diff
@@ -10285,14 +10353,14 @@ func TestValidateNode(t *testing.T) {
-				"spec.taints[0].key":    true,
-				"spec.taints[0].value":  true,
-				"spec.taints[0].effect": true,
+				"metadata.name":                                                                                               true,
```
gofmt changes should not be part of this PR. Do you use go 1.11?

mittachaitu replied:
Yes, I am using  the same go version

jsafrane commented on pkg/apis/core/types.go:
```diff
@@ -650,6 +650,13 @@ type ISCSIVolumeSource struct {
+	// If the ReplacementTImeout value is less than 0, IO will remain queued until the session
+	// is logged back in, or until the user runs the logout command.
+	// +optional
+	ReplacementTimeout string
```
It should be `*int32` (or `*int64`) - the field needs to be optional and nil value should mean the behavior we have today.

prateekpandey14 replied:
@jsafrane just to add what @mittachaitu said , the other thing is if we use integer value, we have to convert it to string before passing it to UpdateISCSI method as iscsi-update operator accepts string args as a value, so anyway we have to do type conversion.

jsafrane replied:
This API is targeted to users, not iscsiadm. Users must see that an integer is expected. You'll get validation for free. Using OpenAPI, the validation can be even done on the client side.

How is the parameter converted internally and passed to iscsiadm is implementation detail.

prateekpandey14 replied:
Agreed what you said, thanks,  we have changed to `*int32` and testing the changes right now.

jsafrane commented on pkg/apis/core/types.go:
```diff
@@ -650,6 +650,13 @@ type ISCSIVolumeSource struct {
+	// iSCSI Replacement Timeout field required to set the iSCSI TimeoutSetting to wait for session re-establishment
+	// before failing SCSI commands back to the application when running the Linux SCSI Layer error handler.
+	// If the ReplacementTimeout value is 0, IO will be failed immediately.
+	// If the ReplacementTImeout value is less than 0, IO will remain queued until the session
```
You validate that the number cannot be negative and this comment allows negative value.

jsafrane commented on pkg/apis/core/types.go:
```diff
@@ -650,6 +650,13 @@ type ISCSIVolumeSource struct {
 	// <target portal>:<volume name> will be created for the connection.
 	// +optional
 	InitiatorName *string
+	// iSCSI Replacement Timeout field required to set the iSCSI TimeoutSetting to wait for session re-establishment
```
Better "ReplacementTimeout controls how long the iSCSI layer should wait for a timed-out path/session to reestablish itself before failing any commands on it."

jsafrane commented on pkg/apis/core/types.go:
```diff
@@ -650,6 +650,13 @@ type ISCSIVolumeSource struct {
 	InitiatorName *string
+	// iSCSI Replacement Timeout field required to set the iSCSI TimeoutSetting to wait for session re-establishment
+	// before failing SCSI commands back to the application when running the Linux SCSI Layer error handler.
+	// If the ReplacementTimeout value is 0, IO will be failed immediately.
```
You should also describe meaning of empty value.

jsafrane commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -1156,6 +1156,13 @@ type ISCSIVolumeSource struct {
+	// If the ReplacementTImeout value is less than 0, IO will remain queued until the session
+	// is logged back in, or until the user runs the logout command.
+	// +optional
+	ReplacementTimeout string `json:"replacementTimeout,omitempty" portobuf:"bytes,13,opt,name=replacementTimeout" protobuf:"varint,13,opt,name=replacementTimeout"`
```
remove `portobuf`.

jsafrane commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -1202,6 +1209,13 @@ type ISCSIPersistentVolumeSource struct {
+	// If the ReplacementTImeout value is less than 0, IO will remain queued until the session
+	// is logged back in, or until the user runs the logout command.
+	// +optional
+	ReplacementTimeout string `json:"replacementTimeout,omitempty" portobuf:"bytes,13,opt,name=replacementTimeout" protobuf:"bytes,13,opt,name=replacementTimeout"`
```
remove `portobuf`.

jsafrane reviewed (commented):
Now it looks much better.

Please remove the first commit and do separate commits for adding the new field (+ its handling) and the generated stuff.

jsafrane commented:
cc @j-griffith @bswartz @redbaron for iSCSI expertise. What do you think about this PR?

bswartz commented:
No objections from me. I don't play around with iSCSI timeout settings personally. It's worth noting that enhancements like this will need to be reimplemented in CSI plugins and/or the common iSCSI library, or they will be lost in the migration to the out-of-tree iSCSI CSI plugin.

Commit:
Add ReplacementTimeout as ISCSIVolumeSource API

  Add ReplacementTimeout field in ISCSIVolumeSource and
  ISCSIPersistentVolumeSource spec as a API changes.
  ReplacementTimeout controls how long the iSCSI layer should wait
  for a timed-out path/session to re-establish itself before failing
  any commands on it.

Signed-off-by: mittachaitu <sai.chaithanya@mayadata.io>

prateekpandey14 commented:
@jsafrane requested changes are done PTAL

jsafrane commented:
/ok-to-test

jsafrane commented on pkg/apis/core/validation/validation.go:
```diff
@@ -707,6 +708,12 @@ func validateISCSIVolumeSource(iscsi *core.ISCSIVolumeSource, fldPath *field.Pat
 	}
+	if iscsi.ReplacementTimeout != nil {
+		timeout := strconv.Itoa(int(*iscsi.ReplacementTimeout))
+		if len(timeout) == 0 {
```
this can't be ever true, `Itoa` never returns empty string

jsafrane commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -1156,6 +1156,15 @@ type ISCSIVolumeSource struct {
+	// If the ReplacementTimeout field is not provided or empty then it remains with
+	// default value i.e 120 seconds.
+	// +optional
+	ReplacementTimeout *int32 `json:"replacementTimeout,omitempty"`
```
Please add protobuf tags.

mittachaitu replied:
protobuf tags are automatically added in autogenerated files i.e in the second commit.

jsafrane commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -1202,6 +1211,15 @@ type ISCSIPersistentVolumeSource struct {
+	// If the ReplacementTimeout field is not provided or empty then it remains with
+	// default value i.e 120 seconds.
+	// +optional
+	ReplacementTimeout *int32 `json:"replacementTimeout,omitempty"`
```
Please add protobuf tags.

mittachaitu replied:
```protobuf``` tag is added in Auto-generated commit.

jsafrane commented on pkg/apis/core/types.go:
```diff
@@ -698,6 +707,15 @@ type ISCSIPersistentVolumeSource struct {
+	// ReplacementTimeout controls how long the iSCSI layer should wait for a
+	// timed-out path/session to reestablish itself before failing any commands on it.
+	// If the ReplacementTimeout value is 0, IO will be failed immediately.
+	// If the ReplacementTImeout value is less than 0, IO will remain queued until the session
```
https://github.com/open-iscsi/open-iscsi does not describe meaning of negative value. How do you know that negative value means basically no timeout?

mittachaitu replied:
@jsafrane https://github.com/open-iscsi/open-iscsi/blob/master/etc/iscsid.conf in Timeout section described about the negative value of replacement timeout.

jsafrane commented on pkg/volume/iscsi/iscsi_util.go:
```diff
@@ -86,10 +91,29 @@ func updateISCSIDiscoverydb(b iscsiDiskMounter, tp string) error {
+//updateISCSINodeTimeouts will update the timeout settings with the
+//values provided in ISCSIVolumeSource and ISCSIPersistentVolumeSource spec
+func updateISCSINodeTimeouts(b iscsiDiskMounter, tp string) {
+	//Updating the iSCSI timeouts settings to keep volume in RW state in the node
```
I think you can remove the comments and add them to documentation of iSCSI driver. It's intended to user and not Kubernetes developer, right?

jsafrane commented on pkg/volume/iscsi/iscsi.go:
```diff
@@ -506,12 +507,23 @@ func getISCSISecretNameAndNamespace(spec *volume.Spec, defaultSecretNamespace st
 }
 
+// updateFromPV will read iSCSI Timeout settings from iSCSI VolumeSource or PersistentVolumeSource
+func getISCSITimeoutSettingInfo(spec *volume.Spec) string {
```
Function name and the comment right above are not in sync.

prateekpandey14 commented:
/test pull-kubernetes-e2e-gce
/test pull-kubernetes-verify
/test pull-kubernetes-bazel-test

prateekpandey14 commented:
/test pull-kubernetes-verify
/test pull-kubernetes-bazel-test

mittachaitu commented:
/test pull-kubernetes-e2e-kops-aws
/test pull-kubernetes-verify

mittachaitu commented:
/test pull-kubernetes-verify

mittachaitu commented:
/ test pull-kubernetes-e2e-gce-100-performance

prateekpandey14 commented:
@jsafrane @saad-ali , as suggested changes are done , can you please approve this PR.

humblec commented:
I have a high level question/thought  on this PR.  That said,  there are many configuration options available for iscsi which is available through its `conf` file. Bringing these parameters to ISCSI spec may not be a good idea. Then what can be brought in is also a question.  IMO, we should consider `target` specific configuration and not really the deamon specific configuration to the spec.  afaict `ReplacementTimeout` does not fall into this  ( `target`) category . It could be configured by admin via conf changes if they need to tweak the default values. There are `many timeout` values available, bringing those to the spec wont be a good idea I believe.  

@jsafrane @saad-ali @rootfs thoughts ?

mittachaitu commented:
> I have a high level question/thought on this PR. That said, there are many configuration options available for iscsi which is available through its `conf` file. Bringing these parameters to ISCSI spec may not be a good idea. Then what can be brought in is also a question. IMO, we should consider `target` specific configuration and not really the deamon specific configuration to the spec. afaict `ReplacementTimeout` does not fall into this ( `target`) category . It could be configured by admin via conf changes if they need to tweak the default values. There are `many timeout` values available, bringing those to the spec wont be a good idea I believe.
> 
> @jsafrane @saad-ali @rootfs thoughts ?

thanks @humblec for your thoughts.
In non k8s environment, the user does iscsi login, iscsi update command (based on his need), does the mount and uses the mountpoint in the application.
In k8s, iSCSI login will be done by kubelet, and by default, settings in iscsi conf file settings will be applied to the session. Any need to modify these settings need re-login which means unmount of volume and restart of the application.
As @bswartz mentioned, this enhancement is worth to be considered for CSI as well.
Regarding your other point, IMO, we should consider the settings which need re-login.

prateekpandey14 commented:
/retest

mittachaitu commented:
/test pull-kubernetes-verify

mittachaitu commented:
/test pull-kubernetes-verify

Commit:
Auto generated files for ISCSIVolumeSource API changes

Signed-off-by: mittachaitu <sai.chaithanya@mayadata.io>

jsafrane commented:
I think that @humblec raised valid points. We can't accommodate every iSCSI option into kubernetes API. And there is a way how to configure the timeout in a config file. Should we document how to change the config instead?
